### PR TITLE
packet: etcd listen on private interface

### PR DESCRIPTION
**Pull request by surajssd**

### CODE REVIEW

**Review by invidian — DISMISSED** [review]
A bit hacky, but looks OK :+1:

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: I would add these `curl` options: `--retry-delay 1 --retry 60 --retry-connrefused --retry-max-time 60 --connect-timeout 20`

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: I would detect if the command was run already and then not touch the file. But if the file was changed, then you need to run `systemctl daemon-reload` to apply the change and also you have to fail the current execution because the service will otherwise use the old environment.

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Added following:

```
          systemctl daemon-reload
          if ! test -f /tmp/etcd-private-interface; then
            touch /tmp/etcd-private-interface
            exit 1
          fi
```

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Nvm

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: done

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: done

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ah, I forgot, maybe it's better to not rely on `exit 1` to restart it after some time but rather directly restart the service:
```
touch /tmp/etcd-private-interface
systemctl daemon-reload
systemctl restart etcd-member &
exit 1
```

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (invidian) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: How about `grep REPLACE_WITH_PRIVATE_IP /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf` instead? With file in `/tmp`, the script will re-run on every boot. I don't think private IP changes on Packet over time.

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: done

**Review by invidian — DISMISSED** [review]

**Comment by pothos** [comment]
Did you try to reach etcd on `127.0.0.1`? Previously it was listening there but now it shouldn't be available on `127.0.0.1` if it is only listening on the private IP.

**Comment by pothos** [comment]
Shouldn't be a problem because we use the private IP everywhere to access it.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: What? Why not use `COREOS_PACKET_IPV4_PRIVATE_0` var? Like [here](https://github.com/kinvolk/lokomotive-kubernetes/blob/ae523cf4df053d6b3cde113b9f947a5d2d10f052/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L96)

You need [this line](https://github.com/kinvolk/lokomotive-kubernetes/blob/ae523cf4df053d6b3cde113b9f947a5d2d10f052/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L70) but... isn't there any reason that would not work?

**Review by rata — CHANGES_REQUESTED** [review]
I think this needs changes. Even if my suggestion doesn't work for some reason, I think this **needs** to explain why the obvious path (using the flatcar metadata information already there) doesn't work.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Nice, but now you need this, I think:
```
[Unit]
Requires=coreos-metadata.service
After=coreos-metadata.service
```

**Review by pothos — CHANGES_REQUESTED** [review]

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (iaguis) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: This is not needed anymore, right?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: This did not work, we might have to move the parameters from env to cmd line flags. That means also provide `ExecStart`.

I also would like to know why etcd service is shipped without any cmd line flags?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Oh yes, this is not needed. But I was testing @rata 's hypothesis and added all the changes in a separate commit.

I will clean it up if what he is saying works. Otherwise we will be back to what existed before.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Not sure I follow, why this didn't work?

And I guess the env vars are the params, right?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Oh, more confused. Does it work or not? I see this comment is more recent that the one that says it doesn't :joy:

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (iaguis) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: With this in `/run/metadata`:

```
COREOS_PACKET_IPV4="1.2.3.4"
```

this doesn't work:

```
[Unit]
Description=Test

[Service]
EnvironmentFile=/run/metadata
Environment="OMG_LISTEN=${COREOS_PACKET_IPV4}"
ExecStart=/bin/bash -c 'env'
```
env shows `OMG_LISTEN=${COREOS_PACKET_IPV4}`

however this works:

```
[Unit]
Description=Test

[Service]
EnvironmentFile=/run/metadata
Environment="OMG_LISTEN=${COREOS_PACKET_IPV4}"
ExecStart=/bin/bash -c 'OMG_LISTEN=${COREOS_PACKET_IPV4} env'
```

env shows `OMG_LISTEN=1.2.3.4`

So it seems systemd only does substitutions inside `ExecStart`.

Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ohhhh, I understand now. Doesn't seem very nice either :-(

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ok so falling back on doing `curl` then?

Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Not sure, my gut feeling is that we can use the flatcar metadata service in a cleaner way than the curl thing. And is more reliable, it retries automatically, for example.

For example, looking quickly at etcd manpage it seems we can also use the parameter `--listen-client-urls` to the command, instead of the env variable, and there we can use this expansion IIUC.

Probably is cleaner to move this from env vars to arguments and use the env var there.

Also, other alternatives might be possible and cleaner than curl too. Like having a systemd bash script that just reads the Flatcar metadata file (using the environment file) and populates the env as this unit is using it. Something like: `printf "ETCD_LISTEN_CLIENT_URLS=\"https://${COREOS_PRIVATE_IP_0}...\"`.

But I think the first option, just moving the env vars to command line arguments (i.e instead of env var `ETCD_LISTEN_CLIENT_URLS` use param `--listen-client-urls` and the same for others, hopefully there is param for all :crossed_fingers: ) seems way cleaner, IMHO.

What do you think?

Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: btw, sysmted documentation on why this is not expanded is here: https://www.freedesktop.org/software/systemd/man/systemd.exec.html

Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Oh, or maybe adding args to the cmdline is not easy because we are using etcd-wrapper, right? If that is not easy to change, I guess I lean towards creating the env file for this systemd unit, then. Like I mentioned previously.

In any case, using flatcar metadata service to get the private IP seems simpler than curl, IMHO.

What do you think?

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: @rata how about now?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Why exit 1 here?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: @surajssd IMHO, it seems a little bit simpler. What do you think?

Not sure if it can be simpler, though. Like an inline systemd unit that has the flatcar metadata environment (so the COREOS_* vars are there) and executes something like:

```
printf "ETCD_LISTEN_CLIENT_URLS=https://%s:2379\n"  $COREOS_PACKET_IPV4_PRIVATE_0
printf "ETCD_LISTEN_PEER_URLS=https://%s:2380\n"    $COREOS_PACKET_IPV4_PRIVATE_0
printf "ETCD_LISTEN_METRICS_URLS=https://%s:2381\n" $COREOS_PACKET_IPV4_PRIVATE_0
```

redirecting the output to a file, and that file is loaded in the environment here? Maybe this intermediate file makes this option not so nice nor readable?

I think creating the environment file that will be used here (in this way or another) seems cleaner than modifying the systemd unit, reload, weird exit values and those. But maybe this is personal taste, no problem :)

**Review by rata — COMMENTED** [review]
I mentioned another option (as this seems quite hacky), but I'm okay with both (this one or another). I trust your decision :)

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: `opt` seems to be collecting random files (bootkube files, persist-data script, now this too). Would it make sense to create a folder there to store this, and avoid a collection of random files? Not sure, just thinking outloud

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: The service needs to be restarted for the new Environment to be used. That's why
```
systemctl restart etcd-member &
exit 1
```
is needed. You could simply use `exit 1` but that would rely on the restarting rules (max retries, and delay) but we want to immediately restart.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: Isn't `EnvironmentFile=/run/metadata/flatcar` needed for the `ExecStartPre` line?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (pothos) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: I would still put this here at the end unless you have already checked that the EnvironmentFile gets reevaluated for each Exec* step:
```
systemctl restart etcd-member &
exit 1
```

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (alban) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: This environment file is sourced in the same unit where it is written (via `ExecStartPre`). Doesn't this make the unit to fail because it will try to read the file before it is written?

According to man systemd.exec(5):
> The files listed with this [EnvironmentFile=] directive will be read shortly before the process is executed (more specifically, after all processes from a previous unit state terminated. This means you can generate these files in one unit state, and read it with this option in the next).

**Comment by alban** [comment]
I also prefer populating a config file rather than updating the systemd service file.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: Sorry that script was not needed.

Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: I am not sure, I understand what you mean here?

The env var file is created in `create-etcd-private-ip-env-file.service` and consumed in `etcd-member.service`. So this works out properly.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: done @rata

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: It is not needed anymore, since it is happening in a separate service.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: It is removed now.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ohhh, Thanks! Not very nice, though :-/

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Thanks!

But please, not do everything, share your opinion too! It's okay to disagree! =)

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]:
```suggestion
        Description=Create a etcd env file to be used by etcd to listen on private interface
```

**Review by rata — COMMENTED** [review]
@surajssd sorry again for all the iterations this went through, and thanks a lot for your patience! :)

Have you tried if this work? I'll try to double check all the systemd dependencies (requires, after, before, etc.) make sense (I don't remember the meaning, will do it later with systemd docs on the next tab :)). Will mark as approve when I do this, just in case =)

This LGTM, though, but want to know what others think :)

**Comment by surajssd** [comment]
> Have you tried if this work? I'll try to double check all the systemd dependencies (requires, after, before, etc.) make sense (I don't remember the meaning, will do it later with systemd docs on the next tab :)). Will mark as approve when I do this, just in case =)

Yes our clusters came up just fine in CI. And in my personal testing as well things were working fine.

**Comment by surajssd** [comment]
I will squash all the changes into one commit with descriptive commit message and put necessary comments.

**Review by invidian — COMMENTED** [review]
LGTM, but other have more experience in that, so I'll let them approve.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: @alban maybe your comment was before the code change to what you can see now? I'm not sure I follow, if not :)

**Review by rata — DISMISSED** [review]
LGTM, But I'd like to understand @alban comment before merging, just in case :-)

nit, but feel free to ignore: I'd explain in the commit why the patch is needed, not how it is done (as that is hopefully clear looking at the diff in this case, I think). Maybe, about the how (as it's not straight forward), I'll explain why using the env var directly didn't work and the env file needs to be created. WDYT?

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (alban) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: Yes, it is fine in this new version :)

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (alban) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Maybe rename `create-etcd-private-ip-env-file.service` to `create-etcd-config.service` since the file name "etcd.config" is generic?

**Review by alban — DISMISSED** [review]
LGTM.

I added a comment about the name "create-etcd-private-ip-env-file.service" but it's up to you.

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:13]: Cool. Will mark as solved, then :)

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:54]: As we have limits on the size of the user-data, I'd consider removing this comment (not sure I would in other circumstances, though). But not sure, your call :-)

Comment (alban) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:54]: This file is converted to json via `ct` and the comment will be deleted in the process before we get the user-data AFAIR

Comment (rata) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:54]: Ohh, makes sense. I've seen comments stayed, but maybe they were comments inside the scripts put here (like here, that is inside inline code https://github.com/kinvolk/lokomotive-kubernetes/pull/161/files#diff-45288c4191a7d4867391fac80437bd5aR216-R219).

Sorry, my bad. Nothing to see here, move along :-P

File: packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: done

**Review by alban — DISMISSED** [review]

**Review by iaguis — DISMISSED** [review]

**Comment by surajssd** [comment]
retriggering tests, since PacketARM failed.

**Review by invidian — APPROVED** [review]
CI is passing.

**Review by pothos — APPROVED** [review]

**Review by alban — APPROVED** [review]

**Pull request merged**
